sdk/core/wasm/src/merkle.rs

```diff
@@ -0,0 +1,181 @@
+use sp_core::{hashing::blake2_256, Hasher};
+use sp_trie::{LayoutV1, MemoryDB, TrieDBMutBuilder, TrieMut};
+use std::collections::hash_map::DefaultHasher;
+
+// ────────────────────────────────────────────────────────────────────────────
+//  NOTE:
+//  The full StorageHub client crate (`client/file-manager`) depends on a large
+//  graph of Substrate-runtime crates that pull in features which cannot be
+//  compiled to `wasm32-unknown-unknown` (e.g. `std` I/O, rocksdb, proc-macro
+//  helpers, etc.).
+//
+//  For the SDK we only need the minimal in-memory Merkle-Patricia trie logic
+//  (`InMemoryFileDataTrie`) to compute Merkle roots client-side.  Therefore we
+//  copy the few small types (ChunkId, ChunkWithId, hashing adapter) instead of
+//  depending on the heavy client crate.  This keeps the WASM package tiny and
+//  avoids compilation failures in browser/Node environments.
+// ────────────────────────────────────────────────────────────────────────────
+
+#[derive(Default, Clone)]
+pub struct Blake2b256Hasher;
+
+impl Hasher for Blake2b256Hasher {
+    type Out = [u8; 32];
+    type StdHasher = DefaultHasher;
+
+    const LENGTH: usize = 32;
+
+    fn hash(data: &[u8]) -> Self::Out {
+        blake2_256(data)
+    }
+}
+
+// Shortcut for the layout we use.
+type TrieLayout = LayoutV1<Blake2b256Hasher>;
+
+type HashOut = <Blake2b256Hasher as Hasher>::Out;
+struct InMemoryFileDataTrie {
+    root: HashOut,
+    memdb: MemoryDB<Blake2b256Hasher>,
+}
+
+impl InMemoryFileDataTrie {
+    fn new() -> Self {
+        let (memdb, root) = MemoryDB::<Blake2b256Hasher>::default_with_root();
+        Self { root, memdb }
+    }
+
+    fn get_root(&self) -> &HashOut {
+        &self.root
+    }
+
+    fn write_chunk(&mut self, chunk_id: ChunkId, data: &Chunk) {
+        let builder = if self.memdb.keys().is_empty() {
+            TrieDBMutBuilder::<TrieLayout>::new(&mut self.memdb, &mut self.root)
+        } else {
+            TrieDBMutBuilder::<TrieLayout>::from_existing(&mut self.memdb, &mut self.root)
+        };
+
+        let mut trie = builder.build();
+        let value = ChunkWithId {
+            chunk_id,
+            data: data.clone(),
+        }
+        .encode();
+        trie.insert(&chunk_id.as_trie_key(), &value)
+            .expect("in-memory insert cannot fail");
+        // committing happens on drop
+    }
+}
+
+// ────────────────────────────────────────────────────────────────────────────
+// WASM‐exposed wrapper for FileTrie
+// ────────────────────────────────────────────────────────────────────────────
+
+#[wasm_bindgen]
+pub struct FileTrie {
+    inner: InMemoryFileDataTrie,
+    next_id: u64,
+}
+
+#[wasm_bindgen]
+impl FileTrie {
+    #[wasm_bindgen(constructor)]
+    pub fn new() -> FileTrie {
+        FileTrie {
+            inner: InMemoryFileDataTrie::new(),
+            next_id: 0,
+        }
+    }
+
+    #[wasm_bindgen]
+    pub fn push_chunk(&mut self, bytes: &[u8]) {
+        assert!(
```

Comment:
Is it really worth a panic here and not just returning an error? I guess the original code is with this assert

Reply:
This is not necessary at all. The inner functions don't have any limitations regarding the size and it will be restricted from the TS part. All the checks will be there!